immuni_exposure_ingestion/models/validators.py

```diff
@@ -0,0 +1,52 @@
+from typing import List
+
+from marshmallow import ValidationError
+from marshmallow.validate import Validator
+
+from immuni_common.models.mongoengine.temporary_exposure_key import TemporaryExposureKey
+from immuni_exposure_ingestion.core import config
+
+
+class TekListValidator(Validator):
+    """
+    A validator for a list of TEKs.
+    """
+
+    def __call__(self, value: List[TemporaryExposureKey]) -> None:
+        """
+        Validations to be performed according to Apple's documentation:
+
+        - Any ENIntervalNumber values from the same user are not unique.
+        - The period of time covered by the data file exceeds 14 days.
+        - There are any gaps in the ENIntervalNumber values for a user.
+        - Any keys in the file have overlapping time windows.
+        - The TEKRollingPeriod value is not 144.
+
+        :param value: the value of the schema being validated.
+        """
+
+        if len(value) == 0:
+            return
+
+        # NOTE: This kind of validation (rolling_period == 144) could be performed at field level.
+        #   We expect all of these validations to change in the future, so having them all together
+        #   here is done on purpose to make things simpler in case of changes in validation.
+        #   We would like to keep field validation more flexible as not to spread strict validation
+        #   logic in many places.
+
+        if any(tek.rolling_period != 144 for tek in value):
+            raise ValidationError("Some rolling values are not 144.")
+
+        if (n_keys := len(value)) > config.MAX_KEYS_PER_UPLOAD:
+            raise ValidationError(
+                f"Too many TEKs. (actual: {n_keys}, max_allowed: {config.MAX_KEYS_PER_UPLOAD})"
+            )
+
+        rolling_start_numbers = set(tek.rolling_start_number for tek in value)
+        min_start_number = min(rolling_start_numbers)
+        expected_start_numbers = set(
+            min_start_number + 144 * i for i in range(config.MAX_KEYS_PER_UPLOAD)
+        )
+
+        if not rolling_start_numbers.issubset(expected_start_numbers):
+            raise ValidationError("Unexpected rolling start numbers identified.")
```

Comment:
I would rather go for a refactoring of these checks to make them more explicit. At the moment, here you are checking for the overlapping periods, as well as eventual repeated rolling start numbers, etc. raising the very same exception.
I would embrace the explicit is better than implicit approach and clearly log why it is failing, as per the following pseudocode:
```
        sorted_teks = sorted(value, key=operator.attrgetter("rolling_start_number"))
        rolling_start_numbers = [t.rolling_start_number for t in sorted_teks]

        if len(rolling_start_numbers) != len(set(rolling_start_numbers)):
            raise ValidationError("Rolling start numbers are not unique")

        next_rolling_start_number = sorted_teks[0].rolling_start_number
        for current in sorted_teks:
            if current.rolling_start_number < next_rolling_start_number:
                raise ValidationError("Overlapping rolling start numbers")
            next_rolling_start_number = current.rolling_start_number + current.rolling_period
```

Reply:
Refactored the implementation, check it out! Also added way more tests